src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -448,10 +449,162 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
 
             ImmutableArray<PendingBranch> pendingReturns = base.Scan(ref badRegion);
             EnforceDoesNotReturn(syntaxOpt: null);
+            enforceMemberNotNull(syntaxOpt: null, this.State);
+
+            foreach (var pendingReturn in pendingReturns)
+            {
+                enforceMemberNotNull(syntaxOpt: pendingReturn.Branch.Syntax, pendingReturn.State);
+
+                if (pendingReturn.Branch is BoundReturnStatement { ExpressionOpt: BoundExpression expr } returnStatement)
+                {
+                    if (pendingReturn.IsConditionalState)
+                    {
+                        if (!IsConstantFalse(expr))
+                        {
+                            // don't check MemberNotWhenTrue state on a 'return false;'
+                            enforceMemberNotNullWhen(returnStatement.Syntax, sense: true, pendingReturn.StateWhenTrue);
+                        }
+                        if (!IsConstantTrue(expr))
+                        {
+                            // don't check MemberNotWhenFalse state on a 'return true;'
+                            enforceMemberNotNullWhen(returnStatement.Syntax, sense: false, pendingReturn.StateWhenFalse);
+                        }
+                    }
```

Comment:
>} [](start = 20, length = 1)

You need an else here. It is possible to return with an unconditional state (in which the variables are definitely assigned).